File: app/src/main/java/edu/cuhk/csci3310/musictheoryapp/Utils/Accidental.java
package edu.cuhk.csci3310.musictheoryapp.Utils;

public enum Accidental
{
    DOUBLE_FLAT(-2, "bb", "bb"), FLAT(-1, "b", "b"), PLAIN(0, "", ""), SHARP(1, "#", "s"), DOUBLE_SHARP(2, "##", "ss"), NATURAL(0, "", "n");

    public final int value;
    public final String string;
    public final String string_for_image;
    Accidental(int value, String string, String string_for_image)
    {
        this.value = value;
        this.string = string;
        this.string_for_image = string_for_image;
    }
}

File: app/src/main/java/com/donald/musictheoryapp/music/musicxml/Accidental.kt
package com.donald.musictheoryapp.music.musicxml

enum class Accidental(
    val xmlValue: String,
    val jsonValue: String
) {

    FlatFlat("flat-flat", "flat-flat"),
    Flat("flat", "flat"),
    Natural("natural", "natural"),
    Sharp("sharp", "sharp"),
    SharpSharp("sharp-sharp", "sharp-sharp");

    val alter: Int
        get() = ordinal - values().size / 2

    companion object {

        fun fromXmlValue(xmlValue: String): Accidental {
            values().forEach { if (it.xmlValue == xmlValue) return it }
            throw IllegalArgumentException()
        }

        fun fromJsonValue(jsonValue: String): Accidental {
            values().forEach { if (it.jsonValue == jsonValue) return it }
            throw IllegalArgumentException()
        }
    }

}